src/test/mina_automation/daemon.ml

```diff
@@ -0,0 +1,183 @@
+      else if retries_remaining > 0 then (
+        Async.printf "Daemon not responding.. retrying (%i/%i)\n"
+          (retry_attempts - retries_remaining)
+          retry_attempts ;
+        let%bind () = after @@ Time.Span.of_sec retry_delay in
+        go (retries_remaining - 1) )
+      else Deferred.Or_error.error_string output
+    in
+    go retry_attempts
+end
+
+module Config = struct
+  module ConfigDirs = struct
+    type t =
+      { root_path : string
+      ; conf : Filename.t
+      ; genesis : Filename.t
+      ; libp2p_keypair : Filename.t
+      }
+
+    let libp2p_keypair_folder t = String.concat [ t.libp2p_keypair; "/privkey" ]
+
+    let create ?(root_path = "/tmp") ?(config_dir = "mina_spun_test")
+        ?(genesis_dir = "mina_genesis_state")
+        ?(p2p_dir = "mina_test_libp2p_keypair") () =
+      (* create empty config dir to avoid any issues with the default config dir *)
+      let conf = Filename.temp_dir ~in_dir:root_path config_dir "" in
+      let genesis = Filename.temp_dir ~in_dir:root_path genesis_dir "" in
+      let libp2p_keypair = Filename.temp_dir ~in_dir:root_path p2p_dir "" in
+      { root_path; conf; genesis; libp2p_keypair }
+
+    let dirs t = [ t.conf; t.genesis; t.libp2p_keypair ]
+
+    let mina_log t = t.conf ^/ "mina.log"
+  end
+
+  type t = { port : int; dirs : ConfigDirs.t }
+
+  let default ?dirs =
+    { port = 3085
+    ; dirs = (match dirs with Some dirs -> dirs | None -> ConfigDirs.create ())
+    }
+
+  let libp2p_keypair_folder t = ConfigDirs.libp2p_keypair_folder t.dirs
+
+  let generate_keys t =
+    let open Deferred.Let_syntax in
+    let%map () =
+      Init.Client.generate_libp2p_keypair_do (libp2p_keypair_folder t) ()
+    in
+    ()
+end
+
+(** 
+  Module [Process] provides functions to interact with a Mina daemon process.
+*)
+module Process = struct
+  type t = { process : Process.t; config : Config.t; client : Client.t }
+
+  (** [force_kill t] sends a kill signal to the process associated with [t] and waits for the process to terminate.
+    @param t The daemon instance containing the process to be killed.
+    @return A deferred result indicating the success or failure of the operation.
+  *)
+  let force_kill t = Utils.force_kill t.process
+end
+
+let archive_blocks t ~archive_address ~(format : Archive_blocks.format) blocks =
+  Executor.run t
+    ~args:
+      ( [ "advanced"
+        ; "archive-blocks"
+        ; "--archive-addres"
+        ; string_of_int archive_address
+        ; "-" ^ Archive_blocks.format_to_string format
+        ]
+      @ blocks )
+
+let dispatch_blocks t ~archive_address ~(format : Archive_blocks.format)
+    ?(sleep = 5) blocks =
+  Deferred.List.iter blocks ~f:(fun block ->
+      Core.Unix.sleep sleep ;
+      archive_blocks t ~archive_address ~format [ block ] () >>| ignore )
+
+let default = Executor.default
+
+let client t ~(config : Config.t) = Client.create ~port:config.port t
+
+let start t (config : Config.t) =
+  let args =
```

Comment:
I'm confused by the CLI args convention. I know we support both, but why we're using `--abc` and `-abc` interchangably here?